### PR TITLE
Implement Flying

### LINKED ISSUE #125: Control Camera instead of Player
For #114 one cool feature would be to be able to control the camera instead of the character with movement keys. Workarounds would be to ~~use / set ClientCheatGhost~~ [impossible due to `#if !UE_BUILD_SHIPPING` around the code], or to manually calculate / update a player's position (setting velocity to 0 in every single frame) to move them around.

### PR DESCRIPTION
**Pull request by LukeSaward1**
Implemented flying since the UE console isn't yet implemented and is suspended indefinitely until we can get working anticheat.
Closes #125

### CODE REVIEW

File: rtil/src/native/character.rs [hunk 1]
```diff
@@ -91,6 +91,19 @@ impl AMyCharacter {
         assert!(!ptr.is_null());
         unsafe { (*ptr).steamid }
     }
+
+    pub fn movement_mode(&mut self) -> u8 {
```
Comment (oberien) [inline rtil/src/native/character.rs:95]: no `mut` needed here

File: tool/prelude.re [hunk 2]
```diff
@@ -35,4 +35,4 @@ fn step_frame(tick_mode: TickMode) {
         let on_tick = comp.on_tick;
         on_tick();
     }
-}
\ No newline at end of file
+}
```
Comment (LukeSaward1) [inline tool/prelude.re:38]: why

File: rtil/src/native/character.rs [hunk 3]
```diff
@@ -91,6 +91,19 @@ impl AMyCharacter {
...
+        self.movement_mut().movement_mode
+    }
+    pub fn set_movement_mode(&mut self, value: u8) {
+        self.movement_mut().movement_mode = value;
+    }
+    pub fn forward_backward_fly_speed(&mut self) -> f32 {
```
Comment (oberien) [inline rtil/src/native/character.rs:101]: no `mut` needed here

File: rtil/src/native/character.rs [hunk 4]
```diff
@@ -91,6 +91,19 @@ impl AMyCharacter {
...
+        self.movement_mut().max_fly_speed
+    }
+    pub fn set_forward_backward_fly_speed(&mut self, value: f32) {
+        self.movement_mut().max_fly_speed = value;
+    }
```
Comment (oberien) [inline rtil/src/native/character.rs:106]: Why did you rename `max_fly_speed` to `forward_backward_fly_speed`? I think keeping with the UE naming helps others understand the code easier.

File: rtil/src/native/character.rs [hunk 5]
```diff
@@ -159,9 +178,12 @@ struct FUniqueNetIdSteam {
 #[rtil_derive::hook_once(AMyCharacter::Tick)]
 fn save(this: usize) {
     CHARACTER.set(this);
-    let my_character = AMyCharacter::get_player();
+    let mut my_character = AMyCharacter::get_player();
```
Comment (oberien) [inline rtil/src/native/character.rs:181]: No `mut` needed here with the above changes.

File: rtil/src/native/character.rs [hunk 6]
```diff
@@ -159,9 +178,12 @@ struct FUniqueNetIdSteam {
...
     log!("Got AMyCharacter: {:#x}", this);
     log!("Got AMyCharacter::RootComponent: {:#x}", my_character.root_component() as *const _ as usize);
     log!("Got AMyCharacter::Controller: {:#x}", my_character.controller() as *const _ as usize);
     log!("Got AMyCharacter::Movement: {:#x}", my_character.movement() as *const _ as usize);
+    log!("Got AMyCharacter::Movement::MovementMode: {:#x}", &mut my_character.movement_mut().movement_mode as *const u8 as usize);
+    log!("Got AMyCharacter::Movement::Acceleration: {:#x}", &mut my_character.movement_mut().acceleration as *const _ as usize);
+    log!("Got AMyCharacter::Movement::MaxFlySpeed : {:#x}", &mut my_character.movement_mut().max_fly_speed as *const _ as usize);
```
Comment (oberien) [inline rtil/src/native/character.rs:188]: No `&mut` required here at all, you can just directly print the `usize` value.

File: tool/movement.re [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+                Text { text: "Swimming" },
+                Text { text: "Flying" },
+            ),
+            selected: movement_mode.number,
+            onchange: fn(index: int) { movement_mode.number = index; },
+        }),
+        UiElement::Button(UiButton {
+            label: Text { text: "Set Movement Mode" },
+            onclick: fn(label: Text) { Tas::set_movement_mode(movement_mode.number); },
+        }),
```
Comment (oberien) [inline tool/movement.re:53]: Does it make sense to have a separate button to set the movement mode instead of just changing it as soon as it's selected?

Comment (oberien) [inline tool/movement.re:53]: I think this makes more sense to have in the `misc` menu as I assume it's more for debugging and doesn't need to keep any state.

File: tool/movement.re [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+        UiElement::Input(Input {
+            label: FLYING_UP_DOWN_VELOCITY_LABEL,
+            input: "300",
```
Comment (oberien) [inline tool/movement.re:56]: It'll always show as `300` when you close and reopen the menu, no matter to which value you set it. You should fetch this from the game itself when the player opens the menu.

File: tool/movement.re [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+                match val {
+                    Result::Ok(val) => { MOVEMENT_STATE.fly_down_up_velocity = val; },
+                    Result::Err(e) => { FLYING_UP_DOWN_VELOCITY_LABEL.text = "Up/Down Flying Velocity [error: invalid input]"; },
+                }
+            },
+            onchange: fn(input: string) {},
+        }),
+        UiElement::Input(Input {
+            label: FLYING_FORWARD_BACKWARD_VELOCITY_LABEL,
+            input: "600",
```
Comment (oberien) [inline tool/movement.re:68]: ~

File: tool/movement.re [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+    id: MOVEMENT_COMPONENT_ID,
+    conflicts_with: List::of(MOVEMENT_COMPONENT_ID),
+    draw_hud: fn(text: string) -> string { text },
+    tick_mode: TickMode::DontCare,
+    requested_delta_time: Option::None,
+    on_tick: fn() {
+        if MOVEMENT_STATE.enable_fly {
+            Tas::set_movement_mode(5);
+            if MOVEMENT_STATE.is_flying_up && !MOVEMENT_STATE.is_flying_down {
+                if Tas::get_movement_mode() == 5 {
```
Comment (oberien) [inline tool/movement.re:95]: This check is a bit redundant as you set the movement_mode above.

File: tool/movement.re [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+                    let vel = Tas::get_velocity();
+                    Tas::set_velocity(Velocity { x: vel.x, y: vel.y, z: (MOVEMENT_STATE.fly_down_up_velocity * -1.) });
+                }
+            } else if !MOVEMENT_STATE.is_flying_up && !MOVEMENT_STATE.is_flying_down {
+                if Tas::get_movement_mode() == 5 {
+                    let vel = Tas::get_velocity();
+                    Tas::set_velocity(Velocity { x: vel.x, y: vel.y, z: 0. });
+                }
+            }
+        }
```
Comment (oberien) [inline tool/movement.re:110]: You should use an enum for the three different cases. It more precisely allows you to express your intent. It also allows you to use `match` on the enum, making the code cleaner. Also you ensure that you are always only ever in one of the three states (currently, what happens when both `is_flying_up` and `is_flying_down` is `true`?).

**Review by oberien — CHANGES_REQUESTED** [review]
Amazing work and already pretty clean work! I just have a few more remarks which should make the code more readable.

File: tool/movement.re [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+        UiElement::Button(UiButton {
+            label: Text { text: "Set Movement Mode" },
+            onclick: fn(label: Text) { Tas::set_movement_mode(movement_mode.number); },
+        }),
+        UiElement::Input(Input {
+            label: FLYING_UP_DOWN_VELOCITY_LABEL,
+            input: "300",
+            onclick: fn(input: string) {
+                let val = input.parse_float();
+                match val {
```
Comment (oberien) [inline tool/movement.re:59]: You could directly `match input.parse_float()`, removing a redundant assignment.

File: tool/movement.re [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+                }
+            },
+            onchange: fn(input: string) {},
+        }),
+        UiElement::Input(Input {
+            label: FLYING_FORWARD_BACKWARD_VELOCITY_LABEL,
+            input: "600",
+            onclick: fn(input: string) {
+                let val = input.parse_float();
+                match val {
```
Comment (oberien) [inline tool/movement.re:71]: You could directly `match input.parse_float()`, removing a redundant assignment.

File: rtil/src/native/character.rs (hunk 3 above)
Comment (LukeSaward1) [reply rtil/src/native/character.rs:101]: Cannot borrow as `mut`.

File: rtil/src/native/character.rs (hunk 5 above)
Comment (LukeSaward1) [reply rtil/src/native/character.rs:181]: Cannot borrow as mut; afaik mut is required.

File: rtil/src/native/character.rs [hunk 14]
```diff
@@ -164,4 +183,7 @@ fn save(this: usize) {
     log!("Got AMyCharacter::RootComponent: {:#x}", my_character.root_component() as *const _ as usize);
     log!("Got AMyCharacter::Controller: {:#x}", my_character.controller() as *const _ as usize);
     log!("Got AMyCharacter::Movement: {:#x}", my_character.movement() as *const _ as usize);
+    log!("Got AMyCharacter::Movement::MovementMode: {:#x}", &my_character.movement().movement_mode as *const u8 as usize);
```
Comment (oberien) [inline rtil/src/native/character.rs:186]: No need for the `u8` cast here, `_` should be enough.

File: tool/misc.re [hunk 15]
```diff
@@ -117,6 +117,30 @@ fn create_misc_menu() -> Ui {
...
+                Text { text: "Navwalking" },
+                Text { text: "Falling" },
+                Text { text: "Swimming" },
+                Text { text: "Flying" },
+            ),
+            selected: Tas::get_movement_mode(),
+            onchange: fn(index: int) { 
+                Tas::set_movement_mode(index);
+                print(f"Movement Mode [mms]: {index}");
+                print(f"Movement Mode [mms]: {Tas::get_movement_mode()}");
```
Comment (oberien) [inline tool/misc.re:141]: Debugging left in?

File: tool/movement.re [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+                MOVEMENT_STATE.enable_fly = false;
+                Tas::set_movement_mode(1);
+            },
+        }),
+        UiElement::Input(Input {
+            label: FLYING_UP_DOWN_VELOCITY_LABEL,
+            input: f"{MOVEMENT_STATE.fly_down_up_velocity}",
+            onclick: fn(input: string) {
+                match input.parse_float() {
+                    Result::Ok(val) => { MOVEMENT_STATE.fly_down_up_velocity = val; },
```
Comment (oberien) [inline tool/movement.re:45]: You need to reset the error text here.

File: tool/movement.re [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+                }
+            },
+            onchange: fn(input: string) {},
+        }),
+        UiElement::Input(Input {
+            label: FLYING_FORWARD_BACKWARD_VELOCITY_LABEL,
+            input: f"{Tas::get_max_fly_speed()}",
+            onclick: fn(input: string) {
+                match input.parse_float() {
+                    Result::Ok(val) => { Tas::set_max_fly_speed(val); },
```
Comment (oberien) [inline tool/movement.re:56]: You need to reset the error text here.

File: tool/movement.re [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+    fly_down_up_velocity: float,
+    fly_state: FlyState,
+}
+
+fn create_movement_menu() -> Ui {
+    Ui::new("Movement:", List::of(
+        UiElement::Button(UiButton {
+            label: Text { text: "Enable Flying" },
+            onclick: fn(label: Text) {
+                MOVEMENT_STATE.enable_fly = true;
```
Comment (oberien) [inline tool/movement.re:27]: I think it makes more sense to set the Tas movement-mode here once instead of every single tick.

File: tool/movement.re [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+        }
+    },
+    on_key_up: fn(key_code: KeyCode) {
+        let key = key_code.to_small();
+        if key == KEY_LEFT_SHIFT.to_small() {
+            MOVEMENT_STATE.fly_state = FlyState::None;
+        } else if key == KEY_SPACE.to_small() {
+            MOVEMENT_STATE.fly_state = FlyState::None;
+        } 
+    },
```
Comment (oberien) [inline tool/movement.re:115]: What if you pressed space, then press shift, then release space a few milliseconds after? That's something that happens pretty regularly when typing, where you still hold the previous character while already having typed the next one.

**Review by oberien — CHANGES_REQUESTED** [review]

**Review by oberien — APPROVED** [review]
Thanks for the PR :+1:
